File: 1035-uncrossed-lines/1035-uncrossed-lines.cpp
class Solution {
public:
    
    map<pair<int,int>,int>m;
    
    int rec(vector<int>&arr1, vector<int>&arr2,int i,int j){
        if(i==arr1.size() || j==arr2.size())
            return 0;
        
        if(m.find({i,j})!=m.end())
            return m[{i,j}];
        
        int temp=0;
        
        if(arr1[i]==arr2[j])
            temp=1+rec(arr1,arr2,i+1,j+1);
        temp=max(temp,rec(arr1,arr2,i+1,j));
        temp=max(temp,rec(arr1,arr2,i,j+1));
        
        return m[{i,j}]= temp;
    }
    
    int maxUncrossedLines(vector<int>& nums1, vector<int>& nums2) {
        return rec(nums1,nums2,0,0);
    }
};

File: 1035-uncrossed-lines/1035-uncrossed-lines.java
class Solution {
    public int maxUncrossedLines(int[] nums1, int[] nums2) {
        int n=nums1.length;
        int m=nums2.length;
        
        int dp[]=new int[m+1];
        
        for(int i=1;i<=n;i++){
            int prev=0;
            for(int j=1;j<=m;j++){
                int temp=dp[j];
                if(nums1[i-1]==nums2[j-1])
                    dp[j]=prev+1;
                else
                    dp[j]=Math.max(dp[j],dp[j-1]);
                prev=temp;
            }
        }
        
        return dp[m];
    }
}